### PR TITLE
[WPT-1763] Add sliding drawer

### PR DESCRIPTION
**Pull request by danrr**
What it says on the tin. Creates a sliding drawer component that comes in from the right side of the screen (left in RTL) with a scrim behind it.

Since this PR is quite big, docs and changelog entry are coming in a separate PR #291.

### CODE REVIEW

**Comment by danrr** [comment]
>These new files do not include the license heading: packages/bpk-component-sliding-drawer/stories.scss

This component is based on `bpk-component-modal` and `packages/bpk-component-modal/stories.scss` doesn't include the license either. It doesn't feel too important but I can add it.

**Comment by k0nserv** [comment]
Hey @danrr could you add a gif or video show casing this component please?

**Comment by danrr** [comment]
[drawer.mov.zip](https://github.com/Skyscanner/backpack/files/1488017/drawer.mov.zip)
 
@k0nserv  here's a video

**Comment by k0nserv** [comment]
drawer

**Comment by k0nserv** [comment]
We just merged a change that we need to have rebased. Please do this before merging this branch

File: .storybook/config.js [hunk 1]
```diff
@@ -51,6 +51,7 @@ configure(() => {
   require('./../packages/bpk-component-mobile-scroll-container/stories');
   require('./../packages/bpk-component-modal/stories');
   require('./../packages/bpk-component-nudger/stories');
+  require('./../packages/bpk-component-sliding-drawer/stories');
```
Comment (matthewdavidson) [inline .storybook/config.js]: Can we position this alphabetically please? (below `bpk-component-slider`)

File: packages/bpk-component-sliding-drawer/package.json [hunk 2]
```diff
@@ -0,0 +1,31 @@
+{
+  "name": "bpk-component-sliding-drawer",
```
Comment (matthewdavidson) [inline packages/bpk-component-sliding-drawer/package.json]: I propose `bpk-component-drawer`, sliding is part of it's behaviour and as such Im not sure it should be in the name. I.e. what happens when we wanna make it fade in as opposed to sliding?

Comment (matthewdavidson) [reply packages/bpk-component-sliding-drawer/package.json]: Perhaps @jamesf3rguson can weigh in.

File: packages/bpk-component-sliding-drawer/src/BpkSlidingDrawerContent.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+import { cssModules } from 'bpk-react-utils';
+import { BpkButtonLink } from 'bpk-component-link';
+import BpkCloseButton from 'bpk-component-close-button';
+
+import STYLES from './bpk-sliding-drawer-content.scss';
+
+const getClassName = cssModules(STYLES);
+
+const BpkSlidingDrawerContent = (props) => {
+  const contentClassNames = [getClassName('bpk-sliding-drawer')];
```
Comment (matthewdavidson) [inline packages/bpk-component-sliding-drawer/src/BpkSlidingDrawerContent.js]: Convention is to first establish all the class name arrays and then do the conditional classes i.e.:
```
const contentClassNames = [getClassName('bpk-sliding-drawer')];
const headerClassNames = [getClassName('bpk-sliding-drawer__heading')];

if (props.className) { contentClassNames.push(props.className); }
if (props.hideTitle) { headerClassNames.push(getClassName('bpk-sliding-drawer__heading--visually-hidden')); }
```

File: packages/bpk-tokens/src/web/base/z-index.json [hunk 4]
```diff
@@ -27,6 +27,9 @@
     },
     "ZINDEX_MODAL": {
       "value": "{!ZINDEX_MODAL}"
+    },
+    "ZINDEX_SLIDING_DRAWER": {
```
Comment (matthewdavidson) [inline packages/bpk-tokens/src/web/base/z-index.json]: This requires a changelog entry.

File: packages/bpk-component-sliding-drawer/src/BpkSlidingDrawerContent.js (hunk 3 above)
Comment (danrr) [reply packages/bpk-component-sliding-drawer/src/BpkSlidingDrawerContent.js]: why? that's just harder to follow

File: .storybook/config.js (hunk 1 above)
Comment (danrr) [reply .storybook/config.js]: It was there from when it was called `overflow-menu`.

My IDE took care of the renaming so it was left where it was.

I'll move it when we settle on a name.

File: packages/bpk-component-sliding-drawer/src/BpkSlidingDrawer.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    super();
+
+    this.state = {
+      show: true,
+    };
+
+    this.onExited = this.onExited.bind(this);
+    this.hide = this.hide.bind(this);
+  }
+  onExited() {
```
Comment (k0nserv) [inline packages/bpk-component-sliding-drawer/src/BpkSlidingDrawer.js]: This should be `onExit` to mess well with the React convention. If we want to express that it's called after exit explicitly we should call i `didExit`

File: packages/bpk-component-sliding-drawer/src/BpkSlidingDrawer.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  render() {
+    const {
+      isOpen, onClose, target, ...rest
+    } = this.props;
+
+    const { show } = this.state;
+
+    delete rest.onClose;
```
Comment (k0nserv) [inline packages/bpk-component-sliding-drawer/src/BpkSlidingDrawer.js]: Isn't this line redundant? There's no way `onClose` can be in rest since we explicitly pick it out on line 50

File: packages/bpk-component-sliding-drawer/src/bpk-sliding-drawer-content.scss [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+ */
+
+@import '~bpk-mixins/index';
+
+.bpk-sliding-drawer {
+  position: fixed;
+  right: 0;
+  z-index: $bpk-zindex-sliding-drawer;
+  width: 90%;
+  max-width: 400 * $bpk-one-pixel-rem;
```
Comment (k0nserv) [inline packages/bpk-component-sliding-drawer/src/bpk-sliding-drawer-content.scss]: This is not a multiple of any of our spacing values atm. Dunno if it should be, but maybe if so make it `17 * $bpk-spacing-base`(408px). 

cc @jamesf3rguson

File: packages/bpk-component-sliding-drawer/src/BpkSlidingDrawer.js (hunk 6 above)
Comment (danrr) [reply packages/bpk-component-sliding-drawer/src/BpkSlidingDrawer.js]: yeah, that's left over, I've removed it, just not updated the PR yet.

File: packages/bpk-component-sliding-drawer/src/BpkSlidingDrawer.js (hunk 5 above)
Comment (danrr) [reply packages/bpk-component-sliding-drawer/src/BpkSlidingDrawer.js]: The callback prop on Transition is called `exited`, figured `onExited` is the best handler name for `exited`

File: packages/bpk-component-sliding-drawer/src/bpk-sliding-drawer-content.scss (hunk 7 above)
Comment (danrr) [reply packages/bpk-component-sliding-drawer/src/bpk-sliding-drawer-content.scss]: spec said `400 * $bpk-one-pixel-rem;` ¯\\\_(ツ)_/¯

Comment (k0nserv) [reply packages/bpk-component-sliding-drawer/src/bpk-sliding-drawer-content.scss]: I'll check with James

**Comment by k0nserv** [comment]
This breaks in various ways on Safari 10 and Safari 11

File: packages/bpk-component-sliding-drawer/src/bpk-sliding-drawer-content.scss (hunk 7 above)
Comment (k0nserv) [reply packages/bpk-component-sliding-drawer/src/bpk-sliding-drawer-content.scss]: Talked with James about it and we should make it a multiple of our smallest spacing value(6px)

Comment (chucklehound) [reply packages/bpk-component-sliding-drawer/src/bpk-sliding-drawer-content.scss]: This seems odd for horizontal elements.  I can understand doing that for vertical rhythm, but it'd be good to understand the reason for doing this horizontally as well.  Why have bpk-one-pixel-rem if everything needs to be a multiple of 6?

File: packages/bpk-component-drawer/src/BpkDrawer.js [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+import BpkDrawerContent from './BpkDrawerContent';
+
+const ScrimBpkDrawerContent = withScrim(BpkDrawerContent);
+
+class BpkDrawer extends Component {
+  constructor() {
+    super();
+
+    this.state = {
+      show: true,
```
Comment (georgegillams) [inline packages/bpk-component-drawer/src/BpkDrawer.js]: This should be provided by the component consumer rather than stored as state inside the component.

File: packages/bpk-component-drawer/src/BpkDrawer.js [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+          onExited={this.onExited}
+          {...rest}
+        />
+      </Portal>
+    );
+  }
+}
+
+BpkDrawer.propTypes = {
+  isOpen: PropTypes.bool.isRequired,
```
Comment (georgegillams) [inline packages/bpk-component-drawer/src/BpkDrawer.js]: I'm unclear as to the difference between the `show` boolean and `isOpen` boolean. Is this one related to whether it is initially shown?

File: packages/bpk-component-drawer/src/BpkDrawer.js (hunk 8 above)
Comment (danrr) [reply packages/bpk-component-drawer/src/BpkDrawer.js]: No, it shouldn't. This state keeps track of the drawer being in view so animations trigger in order. Continued in next comment.

File: packages/bpk-component-drawer/src/BpkDrawer.js (hunk 9 above)
Comment (danrr) [reply packages/bpk-component-drawer/src/BpkDrawer.js]: Yup, a better name would be `isDrawerShown`.

Comment (georgegillams) [reply packages/bpk-component-drawer/src/BpkDrawer.js]: Ah I see - sorry for the misunderstanding. I think `isOpen` would be best as this is consistent with our modals

Comment (danrr) [reply packages/bpk-component-drawer/src/BpkDrawer.js]: @georgegillams I've pushed a commit that hopefully makes things clearer.

Comment (georgegillams) [reply packages/bpk-component-drawer/src/BpkDrawer.js]: Awesome 🎉

File: packages/bpk-component-drawer/src/BpkDrawerContent.js [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+            {props.children}
+          </div>
+        </section>
+        )}
+    </Transition>
+  );
+};
+
+BpkDrawerContent.propTypes = {
+  getDialogRef: PropTypes.func.isRequired,
```
Comment (k0nserv) [inline packages/bpk-component-drawer/src/BpkDrawerContent.js]: We typically don't prefix ref functions with `get`. So this would just be `dialogRef`. Why is it required?

File: packages/bpk-component-drawer/src/BpkDrawerContent.js [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+  closeEvents: PropTypes.shape({
```
Comment (k0nserv) [inline packages/bpk-component-drawer/src/BpkDrawerContent.js]: What's the purpose of these? I don't see them used anywhere.

File: packages/bpk-component-drawer/src/BpkDrawerContent.js (hunk 10 above)
Comment (danrr) [reply packages/bpk-component-drawer/src/BpkDrawerContent.js]: I can rename it but it's something the scrim passes in, so it'll need a bit more work there.

dialogRef is used to give focus to the thing on top of the scrim so that screen readers know what's going on.

Comment (danrr) [reply packages/bpk-component-drawer/src/BpkDrawerContent.js]: If you're really wanting the name changed I can do it in a follow-up PR that touches the scrim, the modal, and this.

File: packages/bpk-component-drawer/src/BpkDrawerContent.js (hunk 11 above)
Comment (danrr) [reply packages/bpk-component-drawer/src/BpkDrawerContent.js]: The `section` has `{...props.closeEvents}`

They stop the drawer from closing if the a touch/mouse event started on the scrim and ended in the content or vice-versa. That was the original behaviour of the modal before it got extracted out.

File: packages/bpk-component-drawer/src/BpkDrawerContent.js (hunk 10 above)
Comment (k0nserv) [reply packages/bpk-component-drawer/src/BpkDrawerContent.js]: Hmm yeah let's defer it. Right I guess it makes sense since this is internal only

File: packages/bpk-component-drawer/src/BpkDrawerContent.js (hunk 11 above)
Comment (k0nserv) [reply packages/bpk-component-drawer/src/BpkDrawerContent.js]: Oaky

**Comment by k0nserv** [comment]
Merging this and we can iterate further on it in more PRs.
